### ISSUE TITLE
StatusException and StatusRuntimeException not serializable

### ISSUE THREAD

**Issue by robeden** [grpc/grpc-java#1913]
StatusException and StatusRuntimeException are not serializable due to containing a Status object (which makes no attempt to be serializable) with no special handling. Since they extend Throwable, they should be serializable.

Here's a snippet from the exception that arises:
```
... snip ...
Caused by: java.io.NotSerializableException: io.grpc.Status
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1184)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
	at java.io.ObjectOutputStream.defaultWriteObject(ObjectOutputStream.java:441)
	at java.lang.Throwable.writeObject(Throwable.java:985)
	at sun.reflect.GeneratedMethodAccessor129.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	... snip ...
```

Our use case is using an RMI-like API to make calls to a server process which is then making gRPC calls. In this case an operation failed and then things go really wrong because we can't see the root problem since the real exception can't be serialized.

I think to be a "good citizen" the Status*Exceptions should be serializable so it would seem that either:

- Status should also be
- The status should be pulled apart and the pieces stored in the exception classes
- Special serialization routines should be put in place to pull the Status apart on write and put them back together on read

If you know which approach you'd like to take, I'd be happy to implement it.

**Comment by ejona86** [comment]
I had been worried about this a bit. I had noticed in passing a little while back. And it seems you will need the serialized version to be stable across versions (unsurprisingly).

Having `Status` be Serializable is probably the proper way to go. It's representing something that is sent over the wire already (albeit without the Throwable cause). We recently added `Metadata` to `StatusException` which is a bit of a can of worms for Serialization, so we probably want to punt and just mark the field transient.

We don't deal with Serialization much, so we'll need to re-read Effective Java to try to remember the pitfalls to avoid. We're also going to want two tests: 1) that makes sure the exception can be serialized and 2) that deserializes an "old" exception to make sure we didn't break compatibility.

Not that it matters, but I'd probably end up re-setting serialVersionUID to 1.

**Comment by carl-mastrangelo** [comment]
Punting this to unscheduled.  We talked about this at length and didn't come up with a good conclusion.

**Comment by robeden** [comment]
Could you summarize the issues?

**Comment by carl-mastrangelo** [comment]
Status *is* the serializable part of the exception.  The exception doesn't really make much sense to serialize, because it is already expressible as a Status.  Status (which is the code and message) is the standardized way of expressing an error (or success).  

A second issue was the possibility of accidentally including the stack trace (which may be private) when encoded.  We feel that it may be better to force users to do this themselves to avoid accidentally.  

It may be possible to mark the status field as `transient`, but again, there needs to be concrete use cases, and generally applicable to multiple users before we can consider this.

**Comment by robeden** [comment]
My argument:
It sounds like you're trying to design around issues with Java itself, which might be a fair thing to but heavily relies on particular usages or non-usages. Since Throwable is Serializable, fulfilling the contract seems appropriate.
For us, gRPC is used inside a framework that reports to another process via serialization. So, we have to crawl through "causes" an strip out gRPC exceptions to make sure problems can be reported.

**Comment by carl-mastrangelo** [comment]
That's one way to look at it, but I don't think it is the right one.  Exceptions are Java specific, and we have an existing and language neutral means of serializing errors (using this term loosely).  C++ doesn't speak Java serialized exceptions.  Neither does Python, Ruby, or any of the other languages.  Having multiple methods of serialization errors is harmful to understanding the gRPC ecosystem.  

Eventually Statuses will be encodable as a [proto](https://github.com/googleapis/googleapis/blob/master/google/rpc/status.proto).  We would prefer there to be one, common, way of serializing errors.  Your use case is uncommon, so you will have to take on the burden of converting it to a serialization format of your choosing.  Is there some reason why this is onerous for you to do?  Shouldn't it be a 20 line helper method to do the conversion for your specific use?

**Comment by robeden** [comment]
I thought we were talking about StatusException and StatusRuntimeException which is part of the grpc-java API. Status doesn't have to change to allow those classes to serialize.

**Comment by carl-mastrangelo** [comment]
SE and SRE are representations of Status.  From your initial comment:

> StatusException and StatusRuntimeException are not serializable due to containing a Status object (which makes no attempt to be serializable) 

Status is not `Serializable` and is unlikely to ever be.  SE and SRE solely exist to extend Exception, and convert a Status into an exception.  Neither of the aforementioned exceptions are really first class, while Status is.  The serialization of Status that the gRPC Java team is willing to support is the Protocol buffer version.   Making SE and SRE `Serialiazable` does not make sense since they were never intended to be, not designed to be part of Java serialization.  The correct way of serializing them is to not do it all, and instead serialize Status as a proto.  

You have to show that making SE and SRE Serializable is generally useful (and not *just* useful to you).  You have to show that it won't complicate the API for marshalling Status (having two ways of doing something complicates the API).  You also have to show that if we did make it serializable, we would not run into API compatibility problems when we wish to modify the internals of those classes.   We are on the hook *for ever* if we do.

**Comment by carl-mastrangelo** [comment]
@louiscryan and @ejona86 who were part of the original discussion may want to chime in.

**Comment by robeden** [comment]
> Status is not Serializable and is unlikely to ever be. SE and SRE solely exist to extend Exception, and convert a Status into an exception.

Nor should it be. Effective Java #75: you should use a custom serialized form anyway. You can easily encode the Status into the official form you want during serialization and I would highly recommend that.

> You have to show that making SE and SRE Serializable is generally useful (and not just useful to you). You have to show that it won't complicate the API for marshalling Status (having two ways of doing something complicates the API). You also have to show that if we did make it serializable, we would not run into API compatibility problems when we wish to modify the internals of those classes. We are on the hook for ever if we do.

Somehow this has become viewed as a personal request. They are **already** Serializable. I did not request it... they are by necessity:
![screen_shot_2017-01-17_at_7_11_34_pm](https://cloud.githubusercontent.com/assets/1891321/22046902/50c626d2-dce9-11e6-930b-140b2a5afd30.png)

I argue this is a bug, not a feature request. It's like throwing an exception when `toString()` is called. It's part of the object contract and is something that shouldn't cause an explosion. You could omit data from the serialized form, but exploding isn't right.

Anyway, I'll bow out now and defer to others...

**Comment by devinrsmith** [comment]
Just ran across this in our own code.

IMO, `SE` and `SRE` should try to be `Serializable` - as it is the contract on `Throwable`. I do notice that `SE` and `SRE` *do* set `serialVersionUID`s, but since `Status` (and `Metadata`) aren't `Serializable`, we get the exception as @robeden original posted.

I think it's completely reasonable to maintain that `Status` and `Metadata` will not (and should not) be `Serializable`. As such, I think it would be best practice if they were marked as `transient` in `SE`/`SRE`.

**Comment by carl-mastrangelo** [comment]
I originally was against the serializability, but as a matter of practicality, I think the perfect may be the enemy of the good.  I would be in favor of making them serializable, even if there may be surprising caveats.

**Comment by cangencer** [comment]
I just ran into this as well and I agree these exceptions need to be serializable and they're currently breaking contract. I would consider this a bug. Marking the `Status` and `Metadata` fields as `transient` should be enough to fix it without any breaking changes.

**Comment by ericmandm** [comment]
Is this topic causing the following issue I am seeing?

When my grpc server creates a new SRE, and adds the causing exception:
`responseObserver.onError(new StatusRuntimeException(Status.INTERNAL.withCause(t)));`

... when the client catches the SRE, the Status.cause field is null.  If I also add a 'withDescription("some text"), the description is available in the client.

My project is using the following versions:
- SpringBoot: 2.1.2-RELEASE
- grpc-spring-boot-starter: 3.1.0 
- grpc-*: 1.24.0

but I also tried these...
- SpringBoot: 2.2.5-RELEASE 
- grpc-spring-boot-starter: 3.2.2
- grpc-*: 1.27.2

**Comment by ejona86** [comment]
@ericmandm, no, your issue is different. The [cause is not transmitted](https://grpc.github.io/grpc-java/javadoc/io/grpc/Status.html#withCause-java.lang.Throwable-) between the server and client.

**Comment by ericmandm** [comment]
Ah Ok, thanks... and sorry I missed that in the documentation.  
Is there a best practice for sending the details about a server side exception information to the client?  For example, is putting json text in the description field, then deserializing at the client end a decent strategy?

**Comment by ejona86** [comment]
@ericmandm, let's move your discussion to https://github.com/grpc/grpc-java/issues/6853

**Comment by sfc-gh-bdeng** [comment]
Reading through this thread and I have a question (probably a dumb one). If StatusException and StatusRuntimeException is not serializable, then when I wrote something like @ericmandm did above

`responseObserver.onError(new StatusRuntimeException(Status.INTERNAL.withCause(t)));`

How is the exception propagated back to the client? In order for the client to get notified there is an exception, it has to be serialized on the server side first, right?

**Comment by ejona86** [comment]
@sfc-gh-bdeng, please open a new issue to discuss that, to avoid confusing this issue. The short answer is this issue is about Java serialization. gRPC [doesn't use](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md) Java serialization for Status. gRPC doesn't transmit the stack trace nor causal chain, the status code is encoded in `grpc-status` header, and the description is in the `grpc-message` header. The exception is just a Java-ism adaptation of the gRPC protocol.

**Comment by sfc-gh-bdeng** [comment]
@ejona86 
Thanks. I have deleted my question and moved it to #10227

Hopefully you can provide some insights when you got some time, especially for my 2nd question on client side exception handling.

**Comment by davidrabinowitz** [comment]
Hi @ejona86 , this issue also [happens in the Google developed Spark BigQuery connector](https://github.com/GoogleCloudDataproc/spark-bigquery-connector/issues/1131). Any upcoming resolution for this?
